src/nbl/ext/ImGui/ImGui.cpp

```diff
@@ -596,106 +924,289 @@ namespace nbl::ext::imgui
+						MDI::SUBALLOCATOR_TRAITS_T::multi_alloc_addr(fillSubAllocator, offsets.size(), offsets.data(), mdiParams.bytesToFill.data(), MDI_ALIGNMENTS.data()); //! (*) we suballocate memory regions from the allocated chunk with required alignment per MDI::E_BUFFER_CONTENT block
+
+						//! linear allocator is used to fill the mdi data within suballocation memory range, 
+						//! there are a few restrictions regarding how MDI::E_BUFFER_CONTENT(s) can be packed,
+						//! and we have really 2 options how data can be allocated:
+						//! - tightly packed data in single request covering all MDI::E_BUFFER_CONTENT within the allocated memory range
+						//! - each of MDI::E_BUFFER_CONTENT allocated separately & each tightly packed
+
+						auto fillDrawBuffers = [&]<MDI::E_BUFFER_CONTENT type>()
+						{
+							const MDI_SIZE_TYPE globalBlockOffset = offsets[type];
+
+							if (globalBlockOffset == INVALID_ADDRESS or mdiBytesFilled[type])
+								return 0u;
+
+							auto* data = mdiData + globalBlockOffset;
+
+							for (int n = 0; n < drawData->CmdListsCount; n++)
+							{
+								auto* cmd_list = drawData->CmdLists[n];
+
+								if constexpr (type == MDI::EBC_INDEX_BUFFERS)
+								{
+									const auto blockStrideToFill = cmd_list->IdxBuffer.Size * sizeof(ImDrawIdx);
+									::memcpy(data, cmd_list->IdxBuffer.Data, blockStrideToFill);
+									data += blockStrideToFill;
+								}
+								else if (type == MDI::EBC_VERTEX_BUFFERS)
+								{
+									const auto blockStrideToFill = cmd_list->VtxBuffer.Size * sizeof(ImDrawVert);
+									::memcpy(data, cmd_list->VtxBuffer.Data, blockStrideToFill);
+									data += blockStrideToFill;
+								}
+							}
+
+							mdiBytesFilled[type] = true;
+							mdiOffsets[type] = globalBlockOffset;
+							return mdiParams.bytesToFill[type];
+						};
+
+						auto fillIndirectStructures = [&]<MDI::E_BUFFER_CONTENT type>()
+						{
+							const MDI_SIZE_TYPE globalBlockOffset = offsets[type];
+
+							if (globalBlockOffset == INVALID_ADDRESS or mdiBytesFilled[type])
+								return 0u;
+
+							auto* const data = mdiData + globalBlockOffset;
+
+							size_t cmdListIndexObjectOffset = {}, cmdListVertexObjectOffset = {}, drawID = {};
+
+							for (int n = 0; n < drawData->CmdListsCount; n++)
+							{
+								const auto* cmd_list = drawData->CmdLists[n];
+								for (int cmd_i = 0; cmd_i < cmd_list->CmdBuffer.Size; cmd_i++)
+								{
+									const auto* pcmd = &cmd_list->CmdBuffer[cmd_i];
+
+									if constexpr (type == MDI::EBC_DRAW_INDIRECT_STRUCTURES)
+									{
+										auto* indirect = reinterpret_cast<VkDrawIndexedIndirectCommand*>(data) + drawID;
+
+										indirect->firstInstance = drawID; // use base instance as draw ID
+										indirect->indexCount = pcmd->ElemCount;
+										indirect->instanceCount = 1u;
+										indirect->firstIndex = pcmd->IdxOffset + cmdListIndexObjectOffset;
+										indirect->vertexOffset = pcmd->VtxOffset + cmdListVertexObjectOffset;
```

Comment:
you can allow for individidual linear allocator allocations for the index and vertex lists by mutating these